src/util/README.md

```diff
@@ -329,3 +329,18 @@ serialisation queries save only that integer hash code of the string.
 
 \subsubsection irep-serialization-ireps Serialization of `irept` Instances
 
+
+\subsection CProver-output CProver output - printing
+Cprover has 3 principal types of output:
+
+* direct output via std::cout & std::cerr (these are in the process of being removed).
+
+* output from invariants / exceptions
+Invariants output to std::cerr - and provide a backtrace and optionally some diagnostic information.
+For more information on invariants, see `invariant.h`
+Exceptions also print to std::cerr - and have a standard 'what()' interface. More information on exceptions can be found in `exception_utils.h`.
```

Comment:
In the case of cbmc_parse_optionst (I haven't looked at all the other drivers, I'd hope they all behave in similar fashion - because... well, why wouldn't they)

• uit::PLAIN
then ui_message_handlert::message_handler
is not left as nullptr but is initialised to 
message_handler = &*console_message_handler;  (with relevant 'flush' type)

for console_message_handlert:
void console_message_handlert::print(
...
  if(level>=4)
  {
    std::cout << message << '\n';
  }
  else
    std::cerr << message << '\n';

• for uit:: XML_UI

void ui_message_handlert::xml_ui_msg(
...
xmlt result;
...
out << result;


• for uit:: JSON_UI
json_stream =
        std::unique_ptr<json_stream_arrayt>(new json_stream_arrayt(out));
where out has been initialised to std::cout
-------------------------------------
So I left that part vague...

Reply:
Hmm, are we trying to document
1) the interface that the developer should use (don't use cout/cerr),
2) the implementation behind any of those interfaces (uses cout/cerr), or
3) the effect (prints something on cout/cerr)?

Reply:
@tautschnig w.r. to exception printing  
This is how I unwrap what's happening. Clarified here, because I _genuinely_ would like to be pointed in the right direction if I got it wrong.
There's a fair bit of unravelling before you actually get to the bottom of it. 
provisos as above - cbmc_parse_options only considered. 

A) 
parse_options.cpp

int parse_options_baset::main()

catch(const invalid_command_line_argument_exceptiont &e)
  {
    **std::cerr << e.what() << "\n";**
    return CPROVER_EXIT_USAGE_ERROR;
  }
  catch(const cprover_exception_baset &e)
  {
    **std::cerr << e.what() << '\n';**
    return CPROVER_EXIT_EXCEPTION;
  }

B)
cbmc_parse_optionst

int cbmc_parse_optionst::get_goto_program(

catch(incorrect_goto_program_exceptiont &e)
  {
    **log.error()** << e.what() << log.eom;
    return CPROVER_EXIT_EXCEPTION;
  }
  catch(const char *e)
  {
    **log.error()** << e << log.eom;
    return CPROVER_EXIT_EXCEPTION;
  }

cbmc_parse_optionst::cbmc_parse_optionst(int argc, const char **argv)
  : parse_options_baset(CBMC_OPTIONS, argc, argv),
    xml_interfacet(cmdline),
    **messaget(ui_message_handler),
    ui_message_handler(cmdline, std::string("CBMC ") + CBMC_VERSION),**
    path_strategy_chooser()
{
}

int get_goto_program_ret =
    get_goto_program(goto_model, options, cmdline, *this, ui_message_handler);

***this** initialises log with cbmc_parse_optionst, which is subclassed from messaget

• The exceptions print to log.error() , i.e. are given a level 1 
then if the output is plain -> 

void console_message_handlert::print(
...
if(level>=4)
{
std::cout << message << '\n';
}
else
**std::cerr << message << '\n';**


So in the case of JSON and XML output - the output will be to std::cout, but otherwise, its out to std::cerr

Reply:
@tautschnig 
Belabour away if you think it'll improve things. We have a common objective :)
I pulled Chris over for a quick review - and that clarified things a little. 

I've rewritten the section to put forward 'what people really _should_ be doing', i.e. using the messaging interface - with a little bit of an introduction to its objectives.

I've kept a subsection for 'how some things are at the moment, but this is _not_ how we should do them going forward'.